### PR TITLE
Article: Hash Tables and Hash Functions

### PR DESCRIPTION
**Pull request by ds-249**
Compiled the first draft on hashing. let me know if I should add anything else.

### CODE REVIEW

File: Hash-Tables-and-Hashing Functions.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
+# Hash Tables and Hashing Functions
+
+## Introduction to hashing
+
+Hashing is designed to solve the problem of needing to efficiently find or store an item in a collection.
```
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: Maybe link to some good external article on what hashing is?

File: Hash-Tables-and-Hashing Functions.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
+# Hash Tables and Hashing Functions
+
+## Introduction to hashing
+
+Hashing is designed to solve the problem of needing to efficiently find or store an item in a collection. 
+For example, if we have a list of 10,000 words of English and we want to check if a given word is in the list, it would be inefficient to successively compare the word with all 10,000 items until we find a match.
```
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: Maybe mention here in passing that even though they are sorted in alphabetical order, like in a dictionary, you still need some time to search for it.

File: Hash-Tables-and-Hashing Functions.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    # Deletion
+    del my_hash_table[key] // throws a ValueError exception if the key is not present
+
+    # Getting all keys and values stored in the dictionary
+
+    keys = my_hash_table.keys()
+    values = my_hash_table.values()
+
+```
```
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: Add a REPL snippet here.

File: Hash-Tables-and-Hashing Functions.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    A.remove(key);
+
+    // Look up
+    myHashTable.get(key); // returns null if the key K is not present
+    myHashTable.containsKey(key); // returns a boolean value, indicating the presence of a key
+    
+    // Number of key, value pairs in the hash table
+    myHashTable.size();
+
+```
```
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: Add a REPL snippet here.

**Comment by alayek** [comment]
In the implementation part, you can make it absolutely clear that some of these languages come with in-built Hashing support.

Like, when you are using Java Hashmap, do mention `java.util.HashMap` in the import.

**Comment by ds-249** [comment]
@alayek I've implemented your suggestions. I've added links to interesting videos on hashing, if that's alright.

**Comment by alayek** [comment]
Yes, add as many good resources you can add. In fact, add a separate section on resources altogether!

File: Hash-Tables-and-Hashing Functions.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+:rocket: [Run Code](https://repl.it/CVt1)
+
+
+## Interesting Links
+
+- If what I've said doesn't make any sense to you, you may want to check out [this video](https://www.youtube.com/watch?v=x05KubVlh_M).
```
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: > If what I've said doesn't make any sense to you

I understand you are striking a humorous tone, but better just say it plainly

> For further reading

File: Hash-Tables-and-Hashing Functions.md (hunk 1 above)
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: I meant to say link to some Wikipedia article that defines Hashing. where you first use the word _Hashing_.

File: Hash-Tables-and-Hashing Functions.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    myHashTable.containsKey(key); // returns a boolean value, indicating the presence of a key
+    
+    // Number of key, value pairs in the hash table
+    myHashTable.size();
+
+```
+:rocket: [Run Code](https://repl.it/CVt1)
+
+
+## Interesting Links
```
Comment (alayek) [inline Hash-Tables-and-Hashing Functions.md]: Call this Section "Resources"

**Comment by alayek** [comment]
@atjonathan @abhisekp @koustuvsinha please check.

File: Hash-Tables-and-Hashing Functions.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+   7                 |    England       |    London
+   8                 |                  | 
+   9                 |                  |
+   10                |                  |
+   11                |  Switzerland     |    Berne
+
+
+Now, in this specific example things work quite well. 
+Our array needs to be big enough to accommodate the longest string, but in this case that’s only 11 slots. 
+And we do waste a bit of space because, for example, there are no 1-letter keys in our data, nor keys between 8 and 10 letters. But in this case, the waste isn’t so bad either. And taking the length of a string is nice and fast, so is the process of finding the value associated with a given key (certainly faster than doing up to five string comparisons).
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~the waste~~ the wasted space

~~And taking~~ Taking

~~so is~~ and so is

File: Hash-Tables-and-Hashing Functions.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+But, what do we do if our dataset has a string which has more than 11 characters? 
+What if we have one another word with 5 characters, "India", and try assigning it to an index using our hash function. Since the index 5 is already occupied, we have to make a call on what to do with it. This is called a collision.
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~index 5~~ 5th index

File: Hash-Tables-and-Hashing Functions.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+If our dataset had a string with thousand characters, and you make an array of thousand indices to store the data, it would result in a wastage of space. If our keys were random words from English, where there are so many words with same length, using length as a hashing function would be fairly useless.
+
+
+## Collision Handling
+
+Two basic methods are used to handle collisions.
+   1. Separate Chaining
+   2. Open Addressing
+
+### Separate Chaining
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~### Separate Chaining~~ #### Separate Chaining

File: Hash-Tables-and-Hashing Functions.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+   8                 |                                    | 
+   9                 |                                    |
+   10                |                                    |
+   11                |    [Switzerland-Berne]             |
+
+
+So to find an item we first go to the bucket then compare keys. This is a popular method, and if link list is used the hash never fills up. The cost for `get(k)` is on average `O(n)` where n is the number of keys in the bucket, total number of keys be N. 
+The problem with separate chaining is that the data structure can grow with out bounds. 
+
+### Open Addressing
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~### Open Addressing~~ #### Open Addressing

File: Hash-Tables-and-Hashing Functions.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+   5                 |    [Spain-Madrid] -> [India-Delhi] |
+   6                 |    [France-Paris]                  |
+   7                 |    [England-London]                |
+   8                 |                                    | 
+   9                 |                                    |
+   10                |                                    |
+   11                |    [Switzerland-Berne]             |
+
+
+So to find an item we first go to the bucket then compare keys. This is a popular method, and if link list is used the hash never fills up. The cost for `get(k)` is on average `O(n)` where n is the number of keys in the bucket, total number of keys be N.
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~So to find an item~~ To find an item, 

~~then compare keys.~~ and then we compare keys.

~~if link list~~ if a list of links

File: Hash-Tables-and-Hashing Functions.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+Methods for Open Addressing
+   - [Linear Probing](https://en.wikipedia.org/wiki/Linear_probing)
+   - [Quadratic Probing](https://en.wikipedia.org/wiki/Quadratic_probing)
+   - [Double Hashing](https://en.wikipedia.org/wiki/Double_hashing)
+
+
+## How to use hashing in your code.
+
+### Python
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~### Python~~ #### Python

File: Hash-Tables-and-Hashing Functions.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    del my_hash_table[key] # throws a ValueError exception if the key is not present
+
+    # Getting all keys and values stored in the dictionary
+    keys = my_hash_table.keys()
+    values = my_hash_table.values()
+
+```
+:rocket: [Run Code](https://repl.it/CVtK)
+
+### Java
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: ~~### Java~~ #### Java

**Comment by ds-249** [comment]
Done.

File: Hash-Tables-and-Hashing Functions.md [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    // Look up
+    myHashTable.get(key); // returns null if the key K is not present
+    myHashTable.containsKey(key); // returns a boolean value, indicating the presence of a key
+    
+    // Number of key, value pairs in the hash table
+    myHashTable.size();
+
+```
+:rocket: [Run Code](https://repl.it/CVt1)
+
```
Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: Remove this line...

Comment (ds-249) [inline Hash-Tables-and-Hashing Functions.md]: which one?

Comment (ds-249) [inline Hash-Tables-and-Hashing Functions.md]: But, @alayek insisted on adding it

Comment (atjonathan) [inline Hash-Tables-and-Hashing Functions.md]: No worries! I trust @alayek :heart:

Comment (ds-249) [inline Hash-Tables-and-Hashing Functions.md]: @atjonathan, you meant the repl link right?

**Comment by atjonathan** [comment]
### **LGTM :+1:**

**Comment by alayek** [comment]
# LGTM :+1:  

Let's :shipit: :package:

**Pull request merged**